File: python/python-sdk/kestrapy/api/invitations_api.py
from typing import Any, List, Optional

from kestrapy.base_api import BaseApi
from kestrapy.models.auth_controller_invitation_user_request import AuthControllerInvitationUserRequest
from kestrapy.models.iam_invitation_controller_api_invitation_create_request import IAMInvitationControllerApiInvitationCreateRequest
from kestrapy.models.iam_invitation_controller_api_invitation_detail import IAMInvitationControllerApiInvitationDetail
from kestrapy.models.invitation import Invitation
from kestrapy.models.invitation_invitation_status import InvitationInvitationStatus
from kestrapy.models.paged_results_iam_invitation_controller_api_invitation_detail import PagedResultsIAMInvitationControllerApiInvitationDetail
from kestrapy.models.query_filter import QueryFilter


class InvitationsApi(BaseApi):

    # ---- Tenant-scoped ----

    def create_invitation(
        self,
        tenant: str,
        request: IAMInvitationControllerApiInvitationCreateRequest,
    ) -> None:
        path = self._tenant_path(tenant, "invitations")
        self._void_request("POST", path, body=request, content_type=self.JSON)

    def list_invitations_by_email(
        self, email: str, tenant: str,
    ) -> List[IAMInvitationControllerApiInvitationDetail]:
        path = self._tenant_path(tenant, "invitations", "email", email)
        return self._json_list_request("GET", path, IAMInvitationControllerApiInvitationDetail)

    def search_invitations(
        self,
        tenant: str,
        email: Optional[str] = None,
        status: Optional[InvitationInvitationStatus] = None,
        page: Optional[int] = None,
        size: Optional[int] = None,
        sort: Optional[List[str]] = None,
        filters: Optional[List[QueryFilter]] = None,
    ) -> PagedResultsIAMInvitationControllerApiInvitationDetail:
        path = self._tenant_path(tenant, "invitations", "search")
        status_val = status.value if status is not None and hasattr(status, 'value') else status
        params = list(self._build_query_params(
            email=email, status=status_val, page=page, size=size,
        ).items())
        self._append_repeated_param(params, "sort", sort)
        self._append_filter_params(params, filters)
        return self._json_request("GET", path, PagedResultsIAMInvitationControllerApiInvitationDetail, params=params)

    def invitation(self, id: str, tenant: str) -> IAMInvitationControllerApiInvitationDetail:
        path = self._tenant_path(tenant, "invitations", id)
        return self._json_request("GET", path, IAMInvitationControllerApiInvitationDetail)

    def delete_invitation(self, id: str, tenant: str) -> None:
        path = self._tenant_path(tenant, "invitations", id)
        self._void_request("DELETE", path)

    # ---- Current user (cross-tenant) ----

    def find_all_invitations_for_current_user(self) -> List[Invitation]:
        path = self._superadmin_path("me", "invitations")
        return self._json_list_request("GET", path, Invitation)

    # ---- Public auth flow (no tenant in path) ----

    def accept_invitation(self, invitation_id: str) -> Any:
        path = self._superadmin_path("invitation", "accept", invitation_id)
        return self._raw_json_request("POST", path)

    def create_from_invitation(
        self, invitation_id: str, request: AuthControllerInvitationUserRequest,
    ) -> Any:
        path = self._superadmin_path("invitation", "create", invitation_id)
        return self._raw_json_request("POST", path, body=request, content_type=self.JSON)

File: go-sdk/kestra_api_client/invitations_api.go
package kestra_api_client

import "context"

type InvitationsAPI struct {
	baseAPI
}

// CreateInvitation creates an invitation. When the invitee already has a user
// account, or when CreateUserIfNotExist is set, the server grants tenant access
// directly and responds 204 No Content — in that case the returned detail is nil
// with a nil error. A genuinely new invitation responds 201 with the detail.
func (a *InvitationsAPI) CreateInvitation(ctx context.Context, tenant string, request IAMInvitationControllerApiInvitationCreateRequest) (*IAMInvitationControllerApiInvitationDetail, error) {
	return doJSON[*IAMInvitationControllerApiInvitationDetail](&a.baseAPI, ctx, "POST", tenantPath(tenant, "invitations"), request, nil)
}

func (a *InvitationsAPI) Invitation(ctx context.Context, id, tenant string) (*IAMInvitationControllerApiInvitationDetail, error) {
	return doJSON[*IAMInvitationControllerApiInvitationDetail](&a.baseAPI, ctx, "GET", tenantPath(tenant, "invitations", id), nil, nil)
}

func (a *InvitationsAPI) ListInvitationsByEmail(ctx context.Context, email, tenant string) ([]IAMInvitationControllerApiInvitationDetail, error) {
	return doJSON[[]IAMInvitationControllerApiInvitationDetail](&a.baseAPI, ctx, "GET", tenantPath(tenant, "invitations", "email", email), nil, nil)
}

func (a *InvitationsAPI) DeleteInvitation(ctx context.Context, id, tenant string) error {
	return a.doVoidJSON(ctx, "DELETE", tenantPath(tenant, "invitations", id), nil, nil)
}

// SearchInvitations lists invitations. The email and status params are honored
// by both Kestra 1.3 and 2.x (2.x marks them deprecated but still maps them
// internally), so prefer them for 1.3+2.x compatibility. The filters array is
// only interpreted by Kestra 2.x — the 1.3 invitation search endpoint has no
// filters parameter, so any filters passed are silently ignored there. Use it
// only for the 2.x-only fields (FilterQuery, FilterExpiredAt, FilterSuperAdmin),
// and don't combine it with email on 2.x (filters take precedence over the
// legacy email mapping there). Mirrors the hybrid shape of SearchApps.
func (a *InvitationsAPI) SearchInvitations(ctx context.Context, tenant string, page, size *int, sort []string, email *string, status *InvitationInvitationStatus, filters []QueryFilter) (*PagedResultsIAMInvitationControllerApiInvitationDetail, error) {
	params := buildQueryParams("page", page, "size", size, "email", email)
	if status != nil {
		// Set explicitly: buildQueryParams' default branch would print a *InvitationInvitationStatus as a pointer address.
		params.Set("status", string(*status))
	}
	appendRepeatedParam(params, "sort", sort)
	appendFilterParams(params, filters)
	return doJSON[*PagedResultsIAMInvitationControllerApiInvitationDetail](&a.baseAPI, ctx, "GET", tenantPath(tenant, "invitations", "search"), nil, params)
}
